File: Week-1/Python/Q3.py
#User function Template for python3

class Solution:
    def count(self, S, m, n): 
        # code here 
        T = [0] * (n + 1)
        T[0] = 1
     
        for i in range(len(S)):
            j = S[i]
            while j <= n:
                T[j] += T[j - S[i]]
                j = j + 1
     
        return T[n]


#{ 
#  Driver Code Starts
#Initial Template for Python 3

import sys
sys.setrecursionlimit(10**6)

if __name__ == '__main__': 
    t = int(input())
    for _ in range(t):
        n,m = list(map(int, input().strip().split()))
        S = list(map(int, input().strip().split()))
        ob = Solution()
        print(ob.count(S,m,n))
# } Driver Code Ends

File: Week-1/Cpp/Q3.cpp
// { Driver Code Starts
#include <bits/stdc++.h>
using namespace std;

 // } Driver Code Ends
class Solution {
  public:
    void filldptable(vector<long long int> &dp, int coin, int amt){
        
        for(int i=coin; i<=amt; i++){
            if(i-coin<0) continue;
            dp[i] = dp[i] + dp[i-coin];
        }
    }
  
  
    long long int count(int S[], int m, int n) {

        // code here.
        vector<long long int> dp(n+1,0);
        dp[0] = 1;
       
        
        for(int i=0; i<m; i++){
            filldptable(dp,S[i],n);
        }
        
        long long int ans=dp[n];
        return ans;
    }
};

// { Driver Code Starts.
int main() {
    int t;
    cin >> t;
    while (t--) {
        int n, m;
        cin >> n >> m;
        int arr[m];
        for (int i = 0; i < m; i++) cin >> arr[i];
        Solution ob;
        cout << ob.count(arr, m, n) << endl;
    }

    return 0;
}  // } Driver Code Ends